### ISSUE TITLE
[Request] Replace `.detach()` with `no_grad()` ctx in `clip_grad_norm_`

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#120611]
Can we replace these _per-gradient_ `.detach()` calls with a `torch.no_grad()` context around the function to reduce CPU overhead?

https://github.com/pytorch/pytorch/blob/b381a4372b627c6179cea1d048e6426dccea9fd5/torch/nn/utils/clip_grad.py#L46
https://github.com/pytorch/pytorch/blob/b381a4372b627c6179cea1d048e6426dccea9fd5/torch/nn/utils/clip_grad.py#L82

The immediate blocker when trying this is circular importing preventing me from using `torch.no_grad()` on `clip_grad_norm_`.

**Comment by awgu** [comment]
cc: @janeyx99

**Comment by janeyx99** [comment]
Taking a step back, should clip_grad_norm ever be run under grad at all?

Also, are some of the detaches are extraneous? For example the one you linked on line 82 is detaching from an already detached grad in grouped_grads, no? Maybe I am not understanding why we detach here.

**Comment by awgu** [comment]
> Taking a step back, should clip_grad_norm ever be run under grad at all?

It should probably not run under grad.

> Also, are some of the detaches are extraneous? For example the one you linked on line 82 is detaching from an already detached grad in grouped_grads, no? Maybe I am not understanding why we detach here.

Good point. From visual inspection, I think the only `.detach()` needed is the one for `.mul_()`.
